File: src/com/sourcegasm/slothvision/control/PID.java
package com.sourcegasm.slothvision.control;

public class PID {
	public static interface Direction {
		public static int NORMAL = 1;
		public static int REVERSED = -1;
	}

	// Some variables
	float p, i, d;
	float err;
	long prevTime = -1;
	float prevErr;
	float integration;
	int direction = 1;

	float imax;

	public PID(float p, float i, float d, float imax, int direction) {
		this.p = p;
		this.i = i;
		this.d = d;
		this.imax = imax;
		this.direction = direction;
	}

	public float calculate(float input, float setpoint) {

		if (prevTime != -1) {

			float dt = (System.nanoTime() - prevTime) / 1000000000.0f;
			prevTime = System.nanoTime();

			float err = setpoint - input;
			float diff = (err - prevErr) / dt;
			integration += err * dt;

			prevErr = err;

			if (integration > imax)
				integration = imax;
			if (integration < -imax)
				integration = -imax;

			return (err * p + integration * i + diff * d) * direction;

		} else {
			prevTime = System.nanoTime();
			return 0;
		}

	}

	public void resetPID() {
		prevErr = 0;
		integration = 0;
		prevTime = -1;
	}
}

File: slothvision/pid/pid.cpp
//
// Created by Vid Drobnič on 9/29/16.
//

#include "pid.h"

PID::PID(double p, double i, double d, double imax, int direction) {
    this->p = p;
    this->i = i;
    this->d = d;
    this->imax = imax;
    this->direction = direction;
}

double PID::calculate(double input, double setpoint) {
    auto now = std::chrono::system_clock::now().time_since_epoch();
    long currentTime = std::chrono::duration_cast<std::chrono::milliseconds>(now).count();

    if (previousTime == -1) {
        previousTime = currentTime;
        return 0;
    }

    double dt = (currentTime - previousTime) / 1000.0;
    previousTime = currentTime;

    double error = setpoint - input;
    double diff = (error - previousError) / dt;
    integration += error * dt;

    previousError = error;

    if (integration > imax) {
        integration = imax;
    } else if (integration < -imax) {
        integration = -imax;
    }

    return (error * p + integration * i + diff * d) * direction;
}

void PID::reset() {
    previousError = 0;
    integration = 0;
    previousTime = -1;
}
